### PR TITLE
[Console] Added support for definition list and horizontal table 

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

I need that in a projet where I want to display some data horizontally.

Usage:
```php
<?php

use Symfony\Component\Console\Helper\TableSeparator;
use Symfony\Component\Console\Input\ArrayInput;
use Symfony\Component\Console\Output\ConsoleOutput;
use Symfony\Component\Console\Style\SymfonyStyle;

require __DIR__.'/vendor/autoload.php';

$io = new SymfonyStyle(new ArrayInput([]), new ConsoleOutput());

$io->table(['a', 'b', 'c'], [[1, 2, 3], [4, 5, 6]]);

$io->table(['a', 'b', 'c'], [[1, 2, 3], [4, 5, 6]], true);

$io->definitionList(
    ['foo' => 'bar'],
    new TableSeparator(),
    'this is a title',
    new TableSeparator(),
    ['foo2' => 'bar2']
);
```

image

### CODE REVIEW

File: src/Symfony/Component/Console/Style/SymfonyStyle.php [hunk 1]
```diff
@@ -185,6 +185,22 @@ public function table(array $headers, array $rows)
...
         $table->setStyle($style);
+        $table->setHorizontal($horizontal);
+
+        $table->render();
+        $this->newLine();
+    }
+
+    public function definitionList(array $list)
+    {
+        $style = clone Table::getStyleDefinition('symfony-style-guide');
```
Comment (Taluu) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: Why the `clone` ?

Comment (lyrixx) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: do not modify the default style. Anyway I copied the code from the "table"

Comment (Taluu) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: Oh it somehow caches the returned style ? In that case, 👍

Comment (Taluu) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: Just checked and indeed. So 👍

**Comment by javiereguiluz** [comment]
I would change the proposed usage by the following:

```php
$io->table(['a', 'b', 'c'], [[1, 2, 3], [4, 5, 6]]);

$io->horizontalTable(['a', 'b', 'c'], [[1, 2, 3], [4, 5, 6]]);

$io->horizontalTable(['id', 'title', 'foo'], [123, 'My title', 'bar']);
```

The boolean param of `table()` is replace by a new method called `horizontalTable()`. The `definitionList()` method is replaced by `horizontalTable()` too because is the same.

What if I have an associative array with the "definition list" information? do this:

```php
$io->horizontalTable(array_keys($foo), array_values($foo));
```

**Comment by lyrixx** [comment]
> The `definitionList()` method is replaced by `horizontalTable()` too because is the same.

Actually, it's ways easier to write:
```php
$io->definitonList(['title' =>  'value', 'title2' => 'value2', /** very long list */]);
```
than
```php
$io->definitonList(['title', 'title2', /* very long list*/], [[1, 2, , /* very long list*/]]);
```
because you will be lost at some point. see [this command](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bundle/FrameworkBundle/Command/AboutCommand.php#L65-L89) for instance

> What if I have an associative array with the "definition list" information? do this:

I did that as the first implementation. But it does not work well. See test to get what I want to achieve :)

File: src/Symfony/Component/Console/CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,7 @@ CHANGELOG
  * added `Question::setTrimmable` default to true to allow the answer to be trimmed
  * added method `preventRedrawFasterThan()` and `forceRedrawSlowerThan()` on `ProgressBar`
  * `Application` implements `ResetInterface`
+ * added support form displaying table horizontally
```
Comment (staabm) [inline src/Symfony/Component/Console/CHANGELOG.md]:
```suggestion
 * added support for displaying table horizontally
```

Comment (OskarStark) [reply src/Symfony/Component/Console/CHANGELOG.md]: ... a table ...

**Review by OskarStark — APPROVED** [review]

**Comment by ro0NL** [comment]
What about 

```php
$output->definitionList(
    ['foo' => 'bar'],
    new TableSeparator(),
    'this is a title',
    new TableSeparator(),
    ['foo2' => 'bar2']
);
```

**Comment by lyrixx** [comment]
It's already supported and tested. Did I miss something?

**Comment by ro0NL** [comment]
> Did I miss something?

Yes, the variadic api :)

i see several advantages;
- duplicate key support (i.e. repetitive sections with the same keys) 
- the expected argument types are clear, array (list), scalar (title) or TableSeparator. In favor of one magic structure

File: src/Symfony/Component/Console/Style/SymfonyStyle.php [hunk 3]
```diff
@@ -185,6 +186,39 @@ public function table(array $headers, array $rows)
...
+        $headers = [];
+        $row = [];
+        foreach ($list as $key => $value) {
+            if ($value instanceof TableSeparator) {
+                $headers[] = $value;
+                $row[] = $value;
+                continue;
+            }
+            if (\is_int($key)) {
+                $headers[] = $value;
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: can we make it span all columns?

**Comment by lyrixx** [comment]
Oh, indeed I miss-read your example :)
This is a good idea, but it requires extra `[``]` :/

**Comment by ro0NL** [comment]
it's still less than the current no. of brackets used https://github.com/symfony/symfony/blob/b2dadc110d4fea28ca981b66aef1d771f238bda5/src/Symfony/Bundle/FrameworkBundle/Command/AboutCommand.php#L64-L90 :} (here we also see multiple `Version` keys btw)

in general i tend to avoid the mixed array structure. The 3rd advantage actually is it supports numeric/integer keys like any other, so less variance.

**Comment by lyrixx** [comment]
@ro0NL Actually, It will be the exact same number of brackets :) 
Anyway, I will try to implement your proposal

File: src/Symfony/Component/Console/Style/SymfonyStyle.php (hunk 3 above)
Comment (lyrixx) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: Yes :) I did it :)

**Comment by lyrixx** [comment]
@ro0NL I finished this PR. I hope you like it :)

File: src/Symfony/Component/Console/Style/SymfonyStyle.php [hunk 4]
```diff
@@ -185,6 +187,42 @@ public function table(array $headers, array $rows)
...
+                $row[] = $value;
+                continue;
+            }
+            if (\is_string($value)) {
+                $headers[] = new TableCell($value, ['colspan' => 2]);
+                $row[] = null;
+                continue;
+            }
+            if (!\is_array($value)) {
+                throw new \InvalidArgumentException('Value should be an array, string or an instance of TableSeparator.');
```
Comment (chalasr) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: `, a string`

File: src/Symfony/Component/Console/Style/SymfonyStyle.php [hunk 5]
```diff
@@ -176,7 +178,7 @@ public function caution($message)
     /**
      * {@inheritdoc}
      */
-    public function table(array $headers, array $rows)
+    public function table(array $headers, array $rows, bool $horizontal = false)
```
Comment (chalasr) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: this breaks inheritance, we need a BC layer

File: src/Symfony/Component/Console/Style/SymfonyStyle.php (hunk 4 above)
Comment (chalasr) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: also let's use the component's `InvalidArgumentException`

**Review by chalasr — CHANGES_REQUESTED** [review]
Nice :)

File: src/Symfony/Component/Console/Style/SymfonyStyle.php (hunk 4 above)
Comment (lyrixx) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: Fixed, thanks

File: src/Symfony/Component/Console/Style/SymfonyStyle.php (hunk 5 above)
Comment (lyrixx) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: I added a new method instead

File: src/Symfony/Component/Console/Style/SymfonyStyle.php [hunk 6]
```diff
@@ -190,6 +193,59 @@ public function table(array $headers, array $rows)
         $this->newLine();
     }
 
+    /**
+     * {@inheritdoc}
```
Comment (chalasr) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: I think we need a new virtual `@method` on `StyleInterface` in order to make this true :)

File: src/Symfony/Component/Console/Style/SymfonyStyle.php [hunk 7]
```diff
@@ -190,6 +193,59 @@ public function table(array $headers, array $rows)
...
+        $table->setHeaders($headers);
+        $table->setRows($rows);
+        $table->setStyle($style);
+        $table->setHorizontal(true);
+
+        $table->render();
+        $this->newLine();
+    }
+
+    public function definitionList(...$list)
```
[the remark this answers is not in the document]
Comment (lyrixx) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: Is it better ?

File: src/Symfony/Component/Console/Helper/Table.php [hunk 8]
```diff
@@ -49,6 +49,8 @@ class Table
      */
     private $rows = [];
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Helper/Table.php]: newline can be removed

File: src/Symfony/Component/Console/Style/SymfonyStyle.php [hunk 9]
```diff
@@ -190,6 +193,67 @@ public function table(array $headers, array $rows)
...
+    }
+
+    /**
+     * Formats a list of key/value horizontally.
+     *
+     * Each row can be one of:
+     * * 'A title'
+     * * ['key' => 'value']
+     * * new TableSeparator()
+     */
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: an `@param` could be `string|array|TableSeparator $list` isnt it?

**Review by ro0NL — APPROVED** [review]

File: src/Symfony/Component/Console/Style/SymfonyStyle.php (hunk 9 above)
Comment (chalasr) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: `@param string|array|TableSeparator ...$list` is the correct one for variadics

Comment (lyrixx) [reply src/Symfony/Component/Console/Style/SymfonyStyle.php]: Added.

File: src/Symfony/Component/Console/Helper/Table.php (hunk 8 above)
Comment (lyrixx) [reply src/Symfony/Component/Console/Helper/Table.php]: Removed.

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Pull request merged**